src/search/utils/collections.h

```diff
@@ -151,6 +152,89 @@ int estimate_unordered_map_bytes(int num_entries) {
     // See comments for _estimate_hash_table_bytes.
     return _estimate_hash_table_bytes<std::unordered_map<Key, Value>>(num_entries);
 }
+
+template<typename T>
+class NamedVector {
+    std::vector<T> elements;
+    std::vector<std::string> names;
+public:
+    NamedVector() = default;
+    NamedVector(NamedVector<T> &&other) = default;
+
+    template<typename ... _Args>
+    void emplace_back(_Args && ... __args) {
+        elements.emplace_back(std::forward<_Args>(__args) ...);
+    }
+
+    void push_back(const T &element) {
+        elements.push_back(element);
+    }
+
+    void set_name(int index, const std::string &name) {
+        if (index >= names.size()) {
+            if (name.empty()) {
+                // All unspecified names are empty by default.
+                return;
+            }
+            names.resize(index + 1, "");
+        }
+        names[index] = name;
+    }
+
+    T &operator[](int index) {
+        return elements[index];
+    }
+
+    std::string get_name(int index) const {
+        if (index < names.size()) {
+            return names[index];
+        } else {
+            // All unspecified names are empty by default.
+            return "";
+        }
+    }
+
+    int size() const {
+        return elements.size();
+    }
+
+    void reserve(int capacity) {
+        elements.reserve(capacity);
+    }
+
+    bool has_names() const {
+        return !names.empty();
+    }
+
+    typename std::vector<T>::reference back() {
+        return elements.back();
+    }
+
+    typename std::vector<T>::iterator begin() {
+        return elements.begin();
+    }
+
+    typename std::vector<T>::iterator end() {
+        return elements.end();
+    }
+
+    typename std::vector<T>::const_iterator begin() const {
+        return elements.begin();
+    }
+
+    typename std::vector<T>::const_iterator end() const {
+        return elements.end();
+    }
+
+    void clear() {
+        elements.clear();
+        names.clear();
+    }
+
+    void resize(int count, T value) {
+        elements.resize(count, value);
```

Comment:
Maybe add a comment here to explain why `names` is not resized. Also, I'd re-order the methods, so that `reserve`, `resize`, `size`, and `clear` are together and `has_names` is together with the other name-related methods.